### PR TITLE
Add codefixer and completion provider to install OpenAPI package from extension methods

### PR DESCRIPTION
**Pull request by captainsafia**
https://github.com/dotnet/aspnetcore/assets/1857993/78ea3d88-cf2d-4abc-96c3-c8df8d69c9c2

### CODE REVIEW

File: src/Framework/AspNetCoreAnalyzers/src/CodeFixes/Dependencies/ThisAndExtensionMethod.cs [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+    {
+        if (obj is ThisAndExtensionMethod other)
+        {
+            return SymbolEqualityComparer.Default.Equals(ThisType, other.ThisType) &&
+                ExtensionMethod.Equals(other.ExtensionMethod, StringComparison.OrdinalIgnoreCase);
+        }
+        return false;
+    }
+
+    public override int GetHashCode() => HashCode.Combine(ThisType, ExtensionMethod);
```
Comment (sebastienros) [inline src/Framework/AspNetCoreAnalyzers/src/CodeFixes/Dependencies/ThisAndExtensionMethod.cs:24]: Should it hash `ExtensionMethod.ToLowerInvariant()` instead since the equality comparison is done with `OrdinalIgnoreCase`?

And I assume that's why it's not a `record struct`?

Comment (captainsafia) [reply src/Framework/AspNetCoreAnalyzers/src/CodeFixes/Dependencies/ThisAndExtensionMethod.cs:24]: Yeah, although I suppose a strict ordinal comparison is the right way to go here given `mapopenapi` and `MapOpenApi` aren't equivalent invocations.

Comment (sebastienros) [reply src/Framework/AspNetCoreAnalyzers/src/CodeFixes/Dependencies/ThisAndExtensionMethod.cs:24]: Now I believe just having a `record struct` might even be better, since:
1- you won't have to write anything, just `internal readonly record struct ThisAndExtensionMethod(ITypeSymbol thisType, string extensionMethod);`
2- it will implement `IEquatable<ThisAndExtensionMethod>` which won't box, and will be used by the dictionary, unlike `Equals(object)` which will box.

**Comment by BrennanConroy** [comment]
This seems like a good general purpose fixer, do we plan on adding more of our methods to this?

File: src/Framework/AspNetCoreAnalyzers/test/Dependencies/AddPackageTests.cs [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+using Microsoft.Extensions.Hosting;
+using Microsoft.Extensions.Logging;
+
+var builder = WebApplication.CreateBuilder(args);
+
+builder.Services.{|CS1061:AddOpenApi|}();
+";
+
+        // Assert
+        await VerifyCS.VerifyCodeFixAsync(source, source);
```
Comment (BrennanConroy) [inline src/Framework/AspNetCoreAnalyzers/test/Dependencies/AddPackageTests.cs:32]: Shouldn't there be a namespace changed when the code fix runs?

**Comment by captainsafia** [comment]
> This seems like a good general purpose fixer, do we plan on adding more of our methods to this?

We could -- there'd be some additional work to do to source all of the APIs that could be invoked here and provide them to this extension point in some way.

For expediency, I just focused on the APIs in the OpenAPI surface area here. I can file an issue to track making this more general purpose.

File: src/Framework/AspNetCoreAnalyzers/test/Dependencies/AddPackageTests.cs (hunk 2 above)
Comment (captainsafia) [reply src/Framework/AspNetCoreAnalyzers/test/Dependencies/AddPackageTests.cs:32]: The code comment on this test outlines why it doesn't actually verify the fix.

TL;DR: Although the AspNetPacakgeInstall API is made available via the external access package, the `PackageInstallService` that actually does the work is internal and ships in a netfx package targeted for consumption in Visual Studio.

I experimented with pulling the `AspNetCoreAddPackageCodeAction.TryCreateCodeActionAsync` invocation into a seperate method that we could mock but that would only allow us to see if we made a call out to the external service, not verify that the actual source changed in the way that was intended.

Comment (captainsafia) [reply src/Framework/AspNetCoreAnalyzers/test/Dependencies/AddPackageTests.cs:37]: I beefed up the tests here so we actually check that the we make a call out to VS to invoke the code action with the correct arguments for the package.

File: src/Framework/AspNetCoreAnalyzers/test/Dependencies/AddPackageTests.cs [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+using Microsoft.Extensions.Hosting;
+using Microsoft.Extensions.Logging;
+
+var builder = WebApplication.CreateBuilder(args);
+
+builder.Services.{|CS1061:AddOpenApi|}();
+";
+
+        // Assert
+        MockAddPackageFixer.Invoked = false;
```
Comment (BrennanConroy) [inline src/Framework/AspNetCoreAnalyzers/test/Dependencies/AddPackageTests.cs:36]: This only reliably works because it's used in the same test class, so tests don't run in parallel right? Should probably mention that as a comment on the class definition so we don't accidentally use it across classes without some thought.

File: src/Framework/AspNetCoreAnalyzers/src/CodeFixes/Dependencies/AddPackageFixer.cs [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+        }
+
+        if (semanticModel == null)
+        {
+            return;
+        }
+
+        var wellKnownTypes = WellKnownTypes.GetOrCreate(semanticModel.Compilation);
+
+        Dictionary<ThisAndExtensionMethod, PackageSourceAndNamespace> _wellKnownExtensionMethodCache = new()
```
Comment (ReubenBond) [inline src/Framework/AspNetCoreAnalyzers/src/CodeFixes/Dependencies/AddPackageFixer.cs:37]: Does `ThisAndExtensionMethod` need to override `Equals` / `GetHashCode` for this to compare symbols correctly (using `SymbolEqualityComparer.Default`)?

File: src/Framework/AspNetCoreAnalyzers/src/CodeFixes/Dependencies/AddPackageFixer.cs [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+                _ => null
+            };
+
+            if (symbolType == null)
+            {
+                return;
+            }
+
+            var targetThisAndExtensionMethod = new ThisAndExtensionMethod(symbolType, methodName);
+            if (_wellKnownExtensionMethodCache.TryGetValue(targetThisAndExtensionMethod, out var packageSourceAndNamespace))
```
Comment (ReubenBond) [inline src/Framework/AspNetCoreAnalyzers/src/CodeFixes/Dependencies/AddPackageFixer.cs:83]: Is a dictionary worthwhile for 2 members, or do we anticipate more in the near future?

File: src/Framework/AspNetCoreAnalyzers/src/CodeFixes/Dependencies/AddPackageFixer.cs [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+
+namespace Microsoft.AspNetCore.Analyzers.Dependencies;
+
+[ExportCodeFixProvider(LanguageNames.CSharp, Name = nameof(AddPackageFixer)), Shared]
+public class AddPackageFixer : CodeFixProvider
+{
+    public override async Task RegisterCodeFixesAsync(CodeFixContext context)
+    {
+        var root = await context.Document.GetSyntaxRootAsync(context.CancellationToken).ConfigureAwait(false);
+        var semanticModel = await context.Document.GetSemanticModelAsync(context.CancellationToken).ConfigureAwait(false);
```
Comment (ReubenBond) [inline src/Framework/AspNetCoreAnalyzers/src/CodeFixes/Dependencies/AddPackageFixer.cs:24]: mega nit, but if there's a decent chance root is null, moving this line after that check avoids any wasted `GetSemanticModelAsync` work.

**Comment by davidfowl** [comment]
> We could -- there'd be some additional work to do to source all of the APIs that could be invoked here and provide them to this extension point in some way.

No could, we should. What is the work to make this work for all of the ASP.NET Core extension methods?

File: src/Framework/AspNetCoreAnalyzers/src/CodeFixes/Dependencies/AddPackageFixer.cs (hunk 5 above)
Comment (captainsafia) [reply src/Framework/AspNetCoreAnalyzers/src/CodeFixes/Dependencies/AddPackageFixer.cs:83]: I anticipate that we'll have a larger collection here (see https://github.com/dotnet/aspnetcore/issues/56092). Whether or not `Dictionary` ends up being the right type for the complete set of extension methods remains to be seen but it seemed sufficient for now.

File: src/Framework/AspNetCoreAnalyzers/src/CodeFixes/Dependencies/AddPackageFixer.cs (hunk 4 above)
Comment (captainsafia) [reply src/Framework/AspNetCoreAnalyzers/src/CodeFixes/Dependencies/AddPackageFixer.cs:37]: Good note -- I reverted this in [Use records struct for comparison](https://github.com/dotnet/aspnetcore/pull/55963/commits/a3531f0d72310eba762d893bb3d69437a555c92b) not recalling the business about `SymbolEqualityComparer`.

**Comment by captainsafia** [comment]
> No could, we should. What is the work to make this work for all of the ASP.NET Core extension methods?

Thanks for the grammar lesson. 🙃 

https://github.com/dotnet/aspnetcore/issues/56092 tracks doing this for everything. Before we do that, we should pilot the package install fixer and a completion provider to recommend extension methods from external packages on primary types with the OpenAPI surface area.

As far as genericizing/automating this, we'll probably need to author some sort of script to scrape the metadata the analyzer consumes from the repo and feed it through.

File: src/Framework/AspNetCoreAnalyzers/test/Dependencies/AddPackageTests.cs (hunk 3 above)
Comment (captainsafia) [reply src/Framework/AspNetCoreAnalyzers/test/Dependencies/AddPackageTests.cs:36]: Added a remark here.

**Comment by davidfowl** [comment]
@IEvangelist Random question, what do the docs do to get this list https://learn.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.builder.iapplicationbuilder?view=aspnetcore-8.0#extension-methods?

I am thinking maybe we can use a similar process to populate the list of methods and their packages.

**Comment by IEvangelist** [comment]
> @IEvangelist Random question, what do the docs do to get this list https://learn.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.builder.iapplicationbuilder?view=aspnetcore-8.0#extension-methods?
> 
> I am thinking maybe we can use a similar process to populate the list of methods and their packages.

I'm not sure how familiar you are with all of this, so I'll speak as if you're not familiar with any of it. This is something that is generated by the DocFx, specifically for the API docs. It uses a process similar to the following:

- Reads a collection NuGet packages
- Generates XML based on these packages
- DocFx reads the XML files and then generates static HTML

But that's something that specific to API docs, not the conceptual content that I believe you're referring to. Do you like the format of the extensions, or what exactly are you thinking? Is the idea to generate reference content for extensions and then map their NuGet packages in the table? Also, I assume you're interested in something like this for .NET Aspire?

File: src/Framework/AspNetCoreAnalyzers/src/CodeFixes/Dependencies/ExtensionMethodsCompletionProvider.cs
Comment (captainsafia) [inline src/Framework/AspNetCoreAnalyzers/src/CodeFixes/Dependencies/ExtensionMethodsCompletionProvider.cs:1]: @JamesNK Can I get your review on this given your experience implementing completion providers in route tooling?

The goal is to add completions for extension methods defined in a separate package then provide an installation codefix for them.

Comment (JamesNK) [reply src/Framework/AspNetCoreAnalyzers/src/CodeFixes/Dependencies/ExtensionMethodsCompletionProvider.cs:1]: It's been a minute since my head was in The Roslyn Zone. I'll do my best.

File: src/Framework/AspNetCoreAnalyzers/src/CodeFixes/Dependencies/AddPackageFixer.cs [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeFixes;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.ExternalAccess.AspNetCore.AddPackage;
+
+namespace Microsoft.AspNetCore.Analyzers.Dependencies;
+
+[ExportCodeFixProvider(LanguageNames.CSharp, Name = nameof(AddPackageFixer)), Shared]
+public class AddPackageFixer : CodeFixProvider
```
Comment (JamesNK) [inline src/Framework/AspNetCoreAnalyzers/src/CodeFixes/Dependencies/AddPackageFixer.cs:19]:
```suggestion
public sealed class AddPackageFixer : CodeFixProvider
```

Comment (JamesNK) [inline src/Framework/AspNetCoreAnalyzers/src/CodeFixes/Dependencies/AddPackageFixer.cs:19]: I know we haven't done this for other fixers, but as someone who is completely new to this code and what it is doing, it would be nice to have a comment summarizing what problem this is solving and how it does it.

File: src/Framework/AspNetCoreAnalyzers/src/CodeFixes/Dependencies/AddPackageFixer.cs [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+        var semanticModel = await context.Document.GetSemanticModelAsync(context.CancellationToken).ConfigureAwait(false);
+        if (semanticModel == null)
+        {
+            return;
+        }
+
+        var wellKnownTypes = WellKnownTypes.GetOrCreate(semanticModel.Compilation);
+        var wellKnownExtensionMethodCache = ExtensionMethodsClass.ConstructFromWellKnownTypes(wellKnownTypes);
+
+        foreach (var diagnostic in context.Diagnostics)
```
Comment (JamesNK) [inline src/Framework/AspNetCoreAnalyzers/src/CodeFixes/Dependencies/AddPackageFixer.cs:38]:
```suggestion
        // Diagnostics are already filtered by FixableDiagnosticIds values.
        foreach (var diagnostic in context.Diagnostics)
```

File: src/Framework/AspNetCoreAnalyzers/src/CodeFixes/Dependencies/AddPackageFixer.cs [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+                    context.RegisterCodeFix(codeAction, diagnostic);
+                }
+            }
+
+        }
+    }
+
+    public sealed override FixAllProvider GetFixAllProvider() => WellKnownFixAllProviders.BatchFixer;
+
+    public override ImmutableArray<string> FixableDiagnosticIds { get; } = ["CS1061"];
```
Comment (JamesNK) [inline src/Framework/AspNetCoreAnalyzers/src/CodeFixes/Dependencies/AddPackageFixer.cs:97]: nit: A comment of what this diagnostic ID is

File: src/Framework/AspNetCoreAnalyzers/src/CodeFixes/Dependencies/ExtensionMethodsCompletionProvider.cs [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Completion;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+
+namespace Microsoft.AspNetCore.Analyzers.RouteEmbeddedLanguage;
+
+[ExportCompletionProvider(nameof(ExtensionMethodsCompletionProvider), LanguageNames.CSharp)]
+[Shared]
+public sealed class ExtensionMethodsCompletionProvider : CompletionProvider
```
Comment (JamesNK) [inline src/Framework/AspNetCoreAnalyzers/src/CodeFixes/Dependencies/ExtensionMethodsCompletionProvider.cs:18]: Add a comment summarizing what the provider's deal is.

Does it replace completion for known types or are values from the completion provider added to the auto-complete? Replacing auto-complete for some types (if that's how it works) feels pretty invasive.

Comment (DamianEdwards) [inline src/Framework/AspNetCoreAnalyzers/src/CodeFixes/Dependencies/ExtensionMethodsCompletionProvider.cs:18]: So excited! 🥳

File: src/Framework/AspNetCoreAnalyzers/src/CodeFixes/Dependencies/AddPackageFixer.cs (hunk 7 above)
Comment (davidfowl) [reply src/Framework/AspNetCoreAnalyzers/src/CodeFixes/Dependencies/AddPackageFixer.cs:19]: You can use copilot chat for that 😁

Comment (captainsafia) [reply src/Framework/AspNetCoreAnalyzers/src/CodeFixes/Dependencies/AddPackageFixer.cs:19]: This can't be sealed since we inherit from it in tests to support mocking out the virtual `TryCreateCodeActionAsync` method to validate that we were able to parse out the symbols and resolve everything correctly from the cache. I'll add a comment about this.

File: src/Framework/AspNetCoreAnalyzers/src/CodeFixes/Dependencies/ExtensionMethodsCompletionProvider.cs (hunk 10 above)
Comment (captainsafia) [reply src/Framework/AspNetCoreAnalyzers/src/CodeFixes/Dependencies/ExtensionMethodsCompletionProvider.cs:25]: The completion provider is additive (assuming I understand the intent of the `context.AddItem` method 😅 ) and is intended to augment the completions list with methods that we think might be helpful. I'll add a comment to this effect.

File: src/Framework/AspNetCoreAnalyzers/src/CodeFixes/Dependencies/AddPackageFixer.cs (hunk 9 above)
Comment (captainsafia) [reply src/Framework/AspNetCoreAnalyzers/src/CodeFixes/Dependencies/AddPackageFixer.cs:110]: Good tip -- Added an example of the emitted diagnostic we target here.

File: src/Framework/AspNetCoreAnalyzers/src/CodeFixes/Dependencies/AddPackageFixer.cs (hunk 7 above)
Comment (captainsafia) [reply src/Framework/AspNetCoreAnalyzers/src/CodeFixes/Dependencies/AddPackageFixer.cs:27]: I added some top-level comments here. Let me know if there's particular LoC that need annotation.

File: src/Framework/AspNetCoreAnalyzers/src/CodeFixes/Dependencies/ExtensionMethodsCompletionProvider.cs (hunk 10 above)
Comment (JamesNK) [reply src/Framework/AspNetCoreAnalyzers/src/CodeFixes/Dependencies/ExtensionMethodsCompletionProvider.cs:25]: What happens if the value is already present in auto-complete? Do items added by this provider look the same or different to other items? Could I see a screenshot?

File: src/Framework/AspNetCoreAnalyzers/src/CodeFixes/Dependencies/ThisAndExtensionMethod.cs [hunk 11]
```diff
@@ -5,4 +5,20 @@
...
+    public override bool Equals(object obj)
+    {
+        return obj is ThisAndExtensionMethod other &&
+            SymbolEqualityComparer.Default.Equals(ThisType, other.ThisType) &&
+            ExtensionMethod == other.ExtensionMethod;
+    }
+
+    public override int GetHashCode()
+    {
+        return HashCode.Combine(ThisType, ExtensionMethod);
```
Comment (ReubenBond) [inline src/Framework/AspNetCoreAnalyzers/src/CodeFixes/Dependencies/ThisAndExtensionMethod.cs:22]: This should use `SymbolEqualityComparer.Default.GetHashCode(ThisType)` to match the above

**Review by ReubenBond — APPROVED** [review]
LGTM. One small change suggested
